include/RAJA/policy/cuda/atomic.hpp

```diff
@@ -43,596 +47,516 @@ namespace RAJA
+  return reinterpret_cast<const unsigned int&>(a) ==
+         reinterpret_cast<const unsigned int&>(b);
 }
-#endif
 
-#if __CUDA_ARCH__ >= 200
-template <typename T>
-RAJA_INLINE __device__ T cuda_atomicMax(T volatile *acc, T value)
+template <typename T,
+          std::enable_if_t<!std::is_same<T, int>::value &&
+                           !std::is_same<T, unsigned int>::value &&
+                           !std::is_same<T, unsigned long long int>::value &&
+                           sizeof(T) == sizeof(unsigned long long int), bool> = true>
+RAJA_INLINE __device__ bool cuda_atomicCAS_equal(const T& a, const T& b)
 {
-  return cuda_atomic_CAS_oper(acc, [=] __device__(T a) {
-    return value > a ? value : a;
-  });
+  return reinterpret_cast<const unsigned long long int&>(a) ==
+         reinterpret_cast<const unsigned long long int&>(b);
 }
 
-// 32-bit signed atomicMax support by CUDA
-template <>
-RAJA_INLINE __device__ int cuda_atomicMax<int>(int volatile *acc,
-                                          int value)
+
+/*!
+ * Generic impementation of any atomic 32-bit or 64-bit operator.
+ * Implementation uses the existing CUDA supplied unsigned 32-bit or 64-bit CAS
+ * operator. Returns the OLD value that was replaced by the result of this
+ * operation.
+ */
+template <typename T, typename OPER>
+RAJA_INLINE __device__ T cuda_atomicCAS(T *acc, OPER&& oper)
 {
-  return ::atomicMax((int *)acc, value);
-}
+  T old = cuda_atomicLoad(acc);
+  T expected;
 
+  do {
+    expected = old;
+    old = cuda_atomicCAS(acc, expected, oper(expected));
+  } while (!cuda_atomicCAS_equal(old, expected));
 
-// 32-bit unsigned atomicMax support by CUDA
-template <>
-RAJA_INLINE __device__ unsigned cuda_atomicMax<unsigned>(unsigned volatile *acc,
-                                                    unsigned value)
-{
-  return ::atomicMax((unsigned *)acc, value);
+  return old;
 }
-#endif
 
-// 64-bit unsigned atomicMax support by CUDA sm_35 and later
-#if __CUDA_ARCH__ >= 350
-template <>
-RAJA_INLINE __device__ unsigned long long cuda_atomicMax<unsigned long long>(
-    unsigned long long volatile *acc,
-    unsigned long long value)
-{
-  return ::atomicMax((unsigned long long *)acc, value);
-}
+
+/*!
+ * Atomic addition
+ */
+using cuda_atomicAdd_builtin_types = list<
+      int
+     ,unsigned int
+     ,unsigned long long int
+     ,float
+#if __CUDA_ARCH__ >= 600
+     ,double
 #endif
+    >;
 
-#if __CUDA_ARCH__ >= 200
-template <typename T>
-RAJA_INLINE __device__ T cuda_atomicInc(T volatile *acc, T val)
+template <typename T,
+          enable_if_is_none_of<T, cuda_atomicAdd_builtin_types>* = nullptr>
+RAJA_INLINE __device__ T cuda_atomicAdd(T *acc, T value)
 {
-  // See:
-  // http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#atomicinc
-  return cuda_atomic_CAS_oper(acc, [=] __device__(T old) {
-    return ((old >= val) ? 0 : (old + 1));
+  return cuda_atomicCAS(acc, [value] __device__(T old) {
```

Comment:
Do you need to put __device__ on a lambda if you're already in a device function?

Reply:
Nope. Removed.